src/main/java/com/conveyal/datatools/manager/models/transform/PreserveCustomFieldsTransformation.java

```diff
@@ -0,0 +1,125 @@
+import com.conveyal.gtfs.loader.Table;
+
+import java.io.*;
+import java.nio.file.*;
+import java.util.*;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.ArrayUtils;
+import org.supercsv.io.CsvMapWriter;
+import org.supercsv.prefs.CsvPreference;
+
+
+/**
+ * This feed transformation will attempt to preserve any custom fields from an entered csv in the final GTFS output.
+ */
+public class PreserveCustomFieldsTransformation extends ZipTransformation {
+    private static List<String> tablePrimaryKeys = new ArrayList<>();
+    private static Table specTable;
+    /** no-arg constructor for de/serialization */
+    public PreserveCustomFieldsTransformation() {}
+
+    public static PreserveCustomFieldsTransformation create(String sourceVersionId, String table) {
+        PreserveCustomFieldsTransformation transformation = new PreserveCustomFieldsTransformation();
+        transformation.sourceVersionId = sourceVersionId;
+        transformation.table = table;
+        return transformation;
+    }
+
+    @Override
+    public void validateParameters(MonitorableJob.Status status) {
+        if (csvData == null) {
+            status.fail("CSV data must not be null (delete table not yet supported)");
+        }
+    }
+
+    /**
+     * This method creates a hash map of the GTFS table keys to the custom CSV values for efficient lookup of custom values.
+     * The hash map key is the key values of the GTFS table (e.g. stop_id for stops) concatenated by an underscore.
+     * The hash map value is the CsvMapReader (mapping of column to row value)
+     */
+    private static HashMap<String, Map<String, String>> createCsvHashMap(CsvMapReader reader, String[] headers) throws IOException {
+        HashMap<String, Map<String, String>> lookup = new HashMap<>();
+        Map<String, String> nextLine;
+        while ((nextLine = reader.read(headers)) != null) {
+            List<String> customCsvKeyValues = tablePrimaryKeys.stream().map(nextLine::get).collect(Collectors.toList());
+            String hashKey = StringUtils.join(customCsvKeyValues, "_");
+            lookup.put(hashKey, nextLine);
+        }
+        return lookup;
+    }
+
+    @Override
+    public void transform(FeedTransformZipTarget zipTarget, MonitorableJob.Status status) {
+        String tableName = table + ".txt";
+        Path targetZipPath = Paths.get(zipTarget.gtfsFile.getAbsolutePath());
+
+        specTable = Arrays.stream(Table.tablesInOrder)
+            .filter(t -> t.name.equals(table))
+            .findFirst()
+            .get();
+        List<String> specTableFields = specTable.specFields().stream().map(f -> f.name).collect(Collectors.toList());
+        tablePrimaryKeys = specTable.getPrimaryKeyNames();
+
+        try(FileSystem targetZipFs = FileSystems.newFileSystem( targetZipPath,(ClassLoader) null)){
```

Comment:
The ClassLoader cast is used because without it the method call is ambiguous. Should I maintain it? 
<img width="446" alt="image" src="https://github.com/ibi-group/datatools-server/assets/63798641/dfc22028-4314-4e12-886f-8cb3c13d493b">

Reply:
What version of the JDK are you using? I don't get this.

Reply:
JDK 18

Reply:
Ignore. I've migrated from JDK 11 to 18.